pkg/api/steve/supportconfigs/generator.go

```diff
@@ -0,0 +1,169 @@
+		ConfigMaps:           scaledContext.Core.ConfigMaps(cspAdapterNamespace),
+		SubjectAccessReviews: scaledContext.K8sClient.AuthorizationV1().SubjectAccessReviews(),
+		ConfigHandler:        supportConfigHandler{},
+	}
+}
+
+func (h *GeneratorHandler) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	err := h.authorize(request)
+	if err != nil {
+		switch err.Error() {
+		case unauthenticatedError:
+			writer.WriteHeader(http.StatusUnauthorized)
+			_, _ = writer.Write([]byte(err.Error()))
+		case forbiddenError:
+			writer.WriteHeader(http.StatusForbidden)
+			_, _ = writer.Write([]byte(err.Error()))
+		default:
+			writer.WriteHeader(http.StatusInternalServerError)
+			logrus.Errorf("Failed to authorize user with error: %s", err.Error())
+		}
+		return
+	}
+	logrus.Infof("Generating supportconfig")
+	archive, err := h.generateSupportConfig()
+	logrus.Infof("Done Generating supportconfig")
+	if err != nil {
+		logrus.Errorf("Error when generating support config %v", err)
+		writer.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	writer.Header().Set("Content-Type", tarContentType)
+	writer.Header().Set("Content-Disposition", "filename=\"supportConfig.tar\"")
+	n, err := io.Copy(writer, archive)
+	logrus.Infof("wrote %v bytes in archive response", n)
+}
+
+// authorize checks to see if the user can get the non-resource url of Endpoint. If not, returns an error, otherwise nil
+func (h *GeneratorHandler) authorize(r *http.Request) error {
+	userInfo, ok := request.UserFrom(r.Context())
+	if !ok {
+		return fmt.Errorf(unauthenticatedError)
+	}
+	response, err := h.SubjectAccessReviews.Create(r.Context(), &authzv1.SubjectAccessReview{
+		Spec: authzv1.SubjectAccessReviewSpec{
+			// This auth method prevents collision with future k8s resources and allows this to be added to custom rules
+			NonResourceAttributes: &authzv1.NonResourceAttributes{
+				Verb: "get",
+				Path: Endpoint,
+			},
+			User:   userInfo.GetName(),
+			Groups: userInfo.GetGroups(),
+			Extra:  convertExtra(userInfo.GetExtra()),
+			UID:    userInfo.GetUID(),
+		},
+	}, metav1.CreateOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to create sar %s", err)
+	}
+	if !response.Status.Allowed {
+		return fmt.Errorf(forbiddenError)
+	}
+	return nil
+}
+
+// generateSupportConfig produces an io.Reader with the supportConfig ready to be returned using a http.ResponseWriter
+// Returns an err if it can't get the support config (note: does not return an err if it can't find the csp config)
+func (h *GeneratorHandler) generateSupportConfig() (io.Reader, error) {
+	mergedConfig, err := h.ConfigHandler.getSupportConfigAsMap()
+	if err != nil {
+		return nil, err
+	}
+	cspConfig, err := h.getCspConfig()
+	// csp adapter won't always be installed
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+	for key, value := range cspConfig {
+		// prefer csp's values - CSP config can see rancher's values and the Cloud Providers
```

Comment:
Can you explain more about this - I see the merge examples in the PR comments but I don't understand functionally what data is being posted to a configmap and what needs to be merged.

Reply:
Here's an overview: 
- There are some values that need to be included in the supportconfig that only the CSP adapter knows. This primarily involves things related to the AWS account that the CSP adapter is installed for (see notes below). When the CSP adapter is not installed, we don't want to use these values. But when they are present, we need to include them.
- In order to support situations where the CSP adapter needs to override a value that is present in the standard supportconfig that rancher produces (such as the support eligible field), this merge process exists to ensure that the values produced by the CSP adapter, which is able to see stuff in k8s and stuff in the cloud provider that it is installed for, are preferred over the ones that rancher produces.

Specific Functionality of the CSP adapter for this support config:
- You can see these fields in code [here](https://github.com/rancher/csp-adapter/blob/88c47fd39dc08ae8daa7d4ecc619cc033f1f9d5a/pkg/manager/types.go). However, keep in mind that from rancher's perspective the CSP adapter can produce anything into this map. Rancher will pick it up and include it while replacing its own values with what is in the map. 
- This includes `SupportEligible bool` - a bool on if this install is eligible for support (always true if CSP adapter is installed)
- Nested Field: CSP - info describing the Cloud Provider setup
  - `Name string` the name of the cloud provider that the csp adapter is using (always `aws` for now)
  - `AcctNumber string` the account number that the CSP adapter is checking for compliance 
- Nested Field: Compliance - info describing the compliance state of rancher
  - `Status string`. If the account is in compliance with the license that it has purchased. Either "Compliant" or "NonCompliant".
  - `Message string`. The message describing why this is in `Status`.

Notes on the CSP adapter:
- The purpose of the CSP adapter (working on better docs, I'm sorry that this isn't better right now) is to track license usage inside of aws. Basically, we want users to be able to purchase support for an arbitrary number of nodes managed by rancher (it's not totally aribitrary but I'll leave those details for the CSP adapter PR). The adapter tracks the usage of these nodes and compares this total to the number that you are entitled to through your license purchases in AWS. If you are using more nodes then you are entitled to, it warns the user and produces output in the supportconfig so that when you open a ticket, support knows if you are outside of compliance. 
- To do this tracking the csp adapter needs to have permissions in the cloud provider. Because of this permissions model and the fact that this functionality is completely optional, we are not including it within core rancher and it is instead a separate chart.

Reply:
> I'm afraid I'm still missing something. Below it looks like we're taking a JSON file from a configmap, converting that JSON file into a tarball, and delivering that as a response. It looks like the content of this JSON file is a few small pieces of information such as the product and platform names.

You've got it. The csp adapter supplies a few other pieces of information (what aws account you are in and if you have gone over on your node allotment), but this is about the scope of it. 

> There is a SUSE-owned tool called [supportconfig](https://github.com/openSUSE/supportutils/blob/master/bin/supportconfig) that does a deep analysis and collection of a host's configuration and logs. This tool is widely used by various support teams. Is this tool at all related to what is being delivered by this endpoint? If so, what component is running the tool? If not, let's please choose a different name for this endpoint.

To answer your questions directly:
- The suse owned tool is not used/ran in any way for this process.
- I think that the idea here was that rancher needs to be able to supply a supportconfig. As you know, rancher operates at a different part of the infra stack than the OS - so we need to collect different data. This endpoint will allow us to produce data that has equivalent relevancy to us as the supportconfig produces for the OS world.
- PM Explicitly calls this a supportconfig. I would prefer to keep the language between the code and the design consistent. Since I can't change the source documents, I think that it is unwise to maintain a separate name in the codebase for what is called for by the design requirements.

Reply:
Could we distill this down into some code comments if possible. Not mandatory since we can always look at PR, but makes reading code in IDE faster.

Reply:
@cbron which parts would you like in the comments? I can't really do a full schema because rancher doesn't know the full schema. I think that the history/naming of the supportconfig might be a bit much. 

I can add: 
- More comments on the fields that rancher does know about (I have some already, if you think they are not sufficient I can add more)
- Comments on on what we expect the supportconfig to do and why we would merge it's results

Does that fit what you were thinking?